vignettes/test-fixtures.Rmd

```diff
@@ -0,0 +1,333 @@
+
+```{r, eval = FALSE}
+local_create_package <- function(dir = file_temp(), env = parent.frame()) {
+  old_project <- proj_get_()
+  
+  # create new folder and package
+  create_package(dir, open = FALSE) # A
+  withr::defer(fs::dir_delete(dir), envir = env) # -A
+  
+  # change working directory
+  setwd(dir) # B
+  withr::defer(setwd(old_project), envir = env) # -B
+  
+  # switch to new usethis project
+  proj_set(dir) # C
+  withr::defer(proj_set(old_project, force = TRUE), envir = env) # -C
+  
+  dir
+}
+```
+
+Note that the cleanup automatically unfolds in the opposite order from the setup. Setup is `A`, then `B`, then `C`; cleanup is `-C`, then `-B`, then `-A`. This is important because we must create directory `dir` before we can make it the working directory; and we must restore the original working directory before we can delete `dir`; we can't delete `dir` while it's still the working directory!
+
+`create_local_package()` is used in over 170 tests. Here's one example that checks that `usethis::use_roxygen_md()` does the setup necessary to use roxygen2 in a package, with markdown support turned on. All 3 expectations consult the DESCRIPTION file, directly or indirectly. So it's very convenient that `create_local_package()` creates a minimal package, with a valid `DESCRIPTION` file, for us to test against. And when the test is done --- poof! --- the package is gone.
+
+```{r eval = FALSE}
+test_that("use_roxygen_md() adds DESCRIPTION fields", {
+  pkg <- local_create_package()
+  use_roxygen_md()
+  
+  expect_true(uses_roxygen_md())
+  expect_equal(desc::desc_get("Roxygen", pkg)[[1]], "list(markdown = TRUE)"))
+  expect_true(desc::desc_has_fields("RoxygenNote", pkg))
+})
+```
+
+## Scope
+
+So far we have applied our test fixture to individual tests, but it's also possible to apply them to a file or package.
+
+### File
+
+If you move the `local_()` call outside of a `test_that()` block, it will affect all tests that come after it. This means that by calling the test fixture at the top of the file you can change the behaviour for all tests. This has both advantages and disadvantages:
+
+-   If you would otherwise have called the fixture in every test, you've saved yourself a bunch of work and duplicate code.
+
+-   But on the downside, if you a test fails and you want to recreate the failure in an interactive environment so you can debug, you need to remember to run all the setup code at the top of the file first.
+
+Generally, I think it's better to copy and paste test fixtures across many tests --- sure, it adds some duplication to your code, but it makes debugging test failures so much easier.
+
+### Package
+
+To run code before any test is run, you can create a file called `test/testthat/setup.R`. If the code in this file needs clean up, you can use the special `teardown_env()`:
+
+```{r, eval = FALSE}
+# Run before any test
+write.csv("mtcars.csv", mtcars)
```

Comment:
Do you even want to show writing a file that's not in the session temp dir 😬?

Reply:
I think it's ok to write into the test directory; I don't really want to write into the session temp dir because there isn't really much point in cleaning it up since the tests are usually run in a fresh session.